language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/CreatedEvent.java

```diff
@@ -27,14 +25,20 @@
 
     private final Optional<Value> contractKey;
 
-    public CreatedEvent(@NonNull List<@NonNull String> witnessParties, @NonNull String eventId, @NonNull Identifier templateId, @NonNull String contractId, @NonNull Record arguments, @NonNull Optional<String> agreementText, @NonNull Optional<Value> contractKey) {
-        this.witnessParties = witnessParties;
+    private final @NonNull Set<@NonNull String> signatories;
+
+    private final @NonNull Set<@NonNull String> observers;
+
+    public CreatedEvent(@NonNull List<@NonNull String> witnessParties, @NonNull String eventId, @NonNull Identifier templateId, @NonNull String contractId, @NonNull Record arguments, @NonNull Optional<String> agreementText, @NonNull Optional<Value> contractKey, @NonNull Set<@NonNull String> signatories, @NonNull Set<@NonNull String> observers) {
+        this.witnessParties = Collections.unmodifiableList(witnessParties);
         this.eventId = eventId;
         this.templateId = templateId;
         this.contractId = contractId;
         this.arguments = arguments;
         this.agreementText = agreementText;
         this.contractKey = contractKey;
+        this.signatories = Collections.unmodifiableSet(signatories);
+        this.observers = Collections.unmodifiableSet(observers);
```

Comment:
Ah, good catch, thanks!

Reply:
I just noticed I'm actually already performing the copy in the `fromProto` method, but the constructor is clearly the safe place where this invariant can be fully protected. I'll move the copies here and add one for the `witnessParties` argument as well.

Reply:
Mh, I was actually making the copy to make the type coming from the ProtoBuffer-generated code (which uses `List`s) compatible with the one I wrote (which uses `Set`s). What if I change the constructor to take a `Collection`? It would be backward compatible.

Reply:
Sorry, I lost the thread here. Which copy are you talking about? In `fromProto`?

Reply:
Sorry, I got it now. Do you want to change the parameter for `witnessParties` to be `Collection` instead of `List`?

Reply:
Mostly the other two to avoid unnecessary extra-copying, but I would add `witnessParties` mostly for consistency. In the end I decided to keep the `witnessParties` as it is and only ask a generic collection for `signatories` and `observers`. Commit coming.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/1814/commits/e7bcb1faf5331d9cdb937c00f8ab7efe71df38c1